language/tools/move-cli/src/sandbox/commands/publish.rs

```diff
@@ -57,18 +60,63 @@ pub fn publish(
         let mut session = vm.new_session(state);
 
         let mut has_error = false;
-        for module in &modules {
-            let mut module_bytes = vec![];
-            module.serialize(&mut module_bytes)?;
+        match override_ordering {
+            None => {
+                for module in &modules {
+                    let mut module_bytes = vec![];
+                    module.serialize(&mut module_bytes)?;
+
+                    let id = module.self_id();
+                    let sender = *id.address();
+
+                    let res =
+                        session.publish_module(module_bytes, sender, &mut gas_status, &log_context);
+                    if let Err(err) = res {
+                        explain_publish_error(err, &state, module)?;
+                        has_error = true;
+                        break;
+                    }
+                }
+            }
+            Some(ordering) => {
+                let module_map: BTreeMap<_, _> = modules
+                    .into_iter()
+                    .map(|m| (m.self_id().name().to_owned(), m))
+                    .collect();
 
-            let id = module.self_id();
-            let sender = *id.address();
+                let mut sender_opt = None;
+                let mut module_bytes_vec = vec![];
+                for name_str in ordering {
+                    let name = IdentStr::new(name_str)?;
+                    match module_map.get(name) {
+                        None => bail!("Invalid module name in publish ordering: {}", name),
+                        Some(module) => {
+                            let mut module_bytes = vec![];
+                            module.serialize(&mut module_bytes)?;
+                            module_bytes_vec.push(module_bytes);
+                            if sender_opt.is_none() {
+                                sender_opt = Some(*module.self_id().address());
+                            }
+                        }
+                    }
+                }
 
-            let res = session.publish_module(module_bytes, sender, &mut gas_status, &log_context);
-            if let Err(err) = res {
-                explain_publish_error(err, &state, module)?;
-                has_error = true;
-                break;
+                match sender_opt {
+                    None => bail!("No modules to publish"),
+                    Some(sender) => {
+                        let res = session.publish_module_bundle(
+                            module_bytes_vec,
+                            sender,
+                            &mut gas_status,
+                            &log_context,
+                        );
+                        if let Err(err) = res {
+                            // TODO (mengxu): explain publish errors in multi-module publishing
+                            println!("Invalid multi-module publishing: {}", err);
```

Comment:
This requires some non-trivial changes to the `explain_publish_error` function. In order not to mess-up with this PR, I prefer to have a separate PR to address this TODO item.